File: exams/aimacana/u3/ExamStrategy/models/quick_sort.py
from typing import List
from .sorting_strategy import SortingStrategy


class QuickSort(SortingStrategy):

    @property
    def name(self) -> str:
        return "QuickSort"

    def sort(self, data: List[int]) -> List[int]:
        arr = list(data)

        def _quick_sort(items: List[int]) -> List[int]:
            if len(items) <= 1:
                return items
            pivot = items[len(items) // 2]
            left = [x for x in items if x < pivot]
            middle = [x for x in items if x == pivot]
            right = [x for x in items if x > pivot]
            return _quick_sort(left) + middle + _quick_sort(right)

        return _quick_sort(arr)

File: exams/chavez/u3/Strategy/strategy/QuickSort.js
const SortingStrategy = require("./SortingStrategy");

class QuickSort extends SortingStrategy {

    sort(array) {

        if (array.length <= 1) {
            return array;
        }

        const pivot = array[array.length - 1];
        const left = [];
        const right = [];

        for (let i = 0; i < array.length - 1; i++) {

            if (array[i] < pivot) {
                left.push(array[i]);
            } else {
                right.push(array[i]);
            }

        }

        return [...this.sort(left), pivot, ...this.sort(right)];
    }

}

module.exports = QuickSort;